packages/peregrine/lib/talons/CheckoutPage/PaymentInformation/usePaymentInformation.js

```diff
@@ -0,0 +1,85 @@
+import { useCallback } from 'react';
+import { useQuery } from '@apollo/react-hooks';
+import { useFieldState } from 'informed';
+
+import { useAppContext } from '../../../context/app';
+import { useCartContext } from '../../../context/cart';
+
+/**
+ *
+ * @param {Function} props.onSave callback to be called when user clicks review order button
+ * @param {Function} props.resetReviewOrderButtonClicked callback to reset the review order button flag
+ * @param {DocumentNode} props.queries.getCheckoutStepQuery query to get the current checkout page step
+ *
+ * @returns {
+ *   doneEditing: Boolean,
+ *   isEditModalHidden: Boolean,
+ *   showEditModal: Function,
+ *   hideEditModal: Function,
+ *   handlePaymentError: Function,
+ *   handlePaymentSuccess: Function,
+ *   currentSelectedPaymentMethod: String,
+ *   checkoutStep: Number,
+ *
+ * }
+ */
+export const usePaymentInformation = props => {
+    const { queries, onSave, resetReviewOrderButtonClicked } = props;
+    const { getCheckoutStepQuery } = queries;
+
+    /**
+     * Definitions
+     */
+
+    const [{ drawer }, { toggleDrawer, closeDrawer }] = useAppContext();
+    const isEditModalActive = drawer === 'edit.payment';
+    const [{ cartId }] = useCartContext();
+    const { value: currentSelectedPaymentMethod } = useFieldState(
+        'selectedPaymentMethod'
+    );
+
+    /**
+     * Query Fetches
+     */
+
+    const { data: checkoutStepData } = useQuery(getCheckoutStepQuery, {
+        variables: cartId
+    });
+
+    const checkoutStep = checkoutStepData
+        ? checkoutStepData.cart.checkoutStep
+        : 1;
+
+    /**
+     * Helper Functions
+     */
+
+    const showEditModal = useCallback(() => {
+        toggleDrawer('edit.payment');
+    }, [toggleDrawer]);
+
+    const hideEditModal = useCallback(() => {
+        closeDrawer('edit.payment');
+    }, [closeDrawer]);
+
+    const handlePaymentSuccess = useCallback(() => {
+        if (onSave) {
+            onSave();
```

Comment:
`onSave` is really just setting the initial checkout step. Use an effect to call it when data has been submitted or on your initial mount after you fetch data and see that data is all present. In this way we can set the step correctly.

See my changes here: https://github.com/magento/pwa-studio/pull/2354/files#diff-bc147cc50ccb76b2ee26bf61bbb89c56R68

Reply:
`onSave` in case of `paymentInformation.js` sets the step to 4. This should only happen after payment success. I understand what you meant by setting the step using `useEffect` and it might work but that seems like a hack. Ideally, the component above should set the step to 3 when it is done and that should be stored in the cache. So even if we refresh the page or perform operations (edit), that should not change.

Reply:
Having `useEffect` might confuse the component in case of authenticated checkout. If the user has a `default` shipping address assigned, and the component fetches it, it might understand that as the presence of data and call the `onSave` function which will set the step to 2 and not perform the mutation of setting shipping address on the cart.

Reply:
You would also do that mutation in an effect. Just solved for this in Shipping Information:

https://github.com/magento/pwa-studio/commit/7396499bb8796b5e5aea0d4e1a22bbecc2880d6a#diff-bc147cc50ccb76b2ee26bf61bbb89c56R103-R116

Reply:
We discussed over chat so logging here:

* Calling the `onSave` or "hasData" callback as it really should be named may be OK so long as the "automatic" flow makes sense for auth.
* Calling `onSave` automatically for payment info in an effect does not make sense. All prior steps are editable except for price adjustments in the "review" step so if we automatically proceed from payment to the review step on a refresh the user will never have a chance to go back if they mess up.